### DISCUSSION TITLE
PROPOSAL: Implement a minimum commission

### DISCUSSION THREAD

**Discussion by Terra854** [crypto-org-chain/chain-main#594] [Ideas] [answered]
Right now, a lot of delegators prefer to stake their tokens to validators that charge 0% commissions. This is not good in the long run as eventually, the staked tokens will be oversaturated around all the 0% validators and also, the operators may have to fork out operating costs from their own pocket.

By implementing a minimum commission percentage (such as what Osmosis has done), validators will be incentivized to run a validator node as they will be able to get additional revenue to cover the costs of running it, potentially allowing them to run the node on better hardware. This will also mitigate the problem where new delegations will go towards the 0% commission validators, spreading out the voting power over all the bonded nodes.

In order to implement it, we will need to perform a network migration to create "a new mainnet" like what Cosmos has done with Stargate (cosmoshub-3 to cosmoshub-4) as there may be some validators have set their max commission percentage to be lower then the proposed minimum commission percentage and the only way to correct it is to edit the resulting genesis file generated by the migration to increase those values

To implement it, these are the steps we could take:

1. Modify the staking module to enforce minimum commissions percentage
2. Create a testnet to test out the migration process
3. When everything is working fine in the testnet, create a software upgrade proposal to implement the changes

When the proposal is approved, the following sequence will occur:

1. The current chain (crypto-org-mainnet-1) halts at a predetermined time or block
2. A new genesis file is generated from the last block, and then edited to increase all validators' commission and max commission percentages to the new minimum if the values are below it and the genesis time to allow time for validators to do the migration.
3. All validators upgrade to the new binary
4. The new chain starts (tentatively named crypto-org-mainnet-2) when the new genesis time has reached

**Comment by samuel-tay** [comment]
I'm not sure if the new genesis file can be generated without the private keys of each individual validator? if there are no changes then I think that's fine, but for those 0% validators and those who have changed their validator settings, then I believe this would be difficult to achieve.

Will delegators need to do anything?

**Reply by Terra854** [reply]
Afaik, there isn't a need for validators to share their private keys as the genesis file only requires their pubkeys which is exported from the last known block before the halt.

Also, it seems like the minimum commission feature is currently being upstreamed into the Cosmos SDK, so depending on how it handles validator's max commission that are below the new minimums, we may no longer need to create a new chain like originally planned

All validators need to do is to upgrade to the new binary when the upgrade time hits

**Comment by tomtau** [comment, marked as answer]
Thank you for the proposal. IMHO the minimum commission is something of interest to potentially any Cosmos SDK-based project. In fact, it is currently being upstreamed to Cosmos SDK:
https://github.com/cosmos/cosmos-sdk/pull/9245

Instead of forking Cosmos SDK, modifying the staking module and starting a new network, it may be preferable to upgrade to the future official Cosmos SDK release that includes these changes for the following reasons:

1. the changes in the upstream Cosmos SDK repository will receive more throughout architecture and code reviews;
2. they will be maintained in the long term (so there is less overhead of maintaining a fork);
3. likely, the release that contains them will also have a state migration code for the upgrade module handler (which will provide a bit smoother process than starting a new network from the exported state).
